### PR TITLE
ROX-9996: Fix image signature verification policy criteria for images verified by multiple integrations

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

Fixes to the augmented object building and to the alert printer.

Split `executeTemplate` function to separately usable parts.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

### CODE REVIEW

File: pkg/booleanpolicy/augmentedobjs/construct.go [hunk 1]
```diff
@@ -236,43 +236,22 @@ func ConstructImage(image *storage.Image) (*pathutil.AugmentedObj, error) {
 	}
 
 	if features.ImageSignatureVerification.Enabled() {
-		i := 0
-		// Since policies query for image verification status as a single boolean field, we add it to the image here.
+		ids := []string{}
```
Comment (md2119) [inline pkg/booleanpolicy/augmentedobjs/construct.go:239]: pre-allocate:
`id := make([]string, 0, len(image.GetSignatureVerificationData().GetResults()))`

File: pkg/booleanpolicy/default_policies_test.go [hunk 2]
```diff
@@ -2218,7 +2222,7 @@ func (suite *DefaultPoliciesTestSuite) TestImageVerified() {
 		{
 			values:          []string{verifier2},
 			negate:          false,
-			expectedMatches: allImages,
+			expectedMatches: allImages.Difference(set.NewFrozenStringSet("verified_by_2_and_3")),
```
Comment (md2119) [inline pkg/booleanpolicy/default_policies_test.go:2225]: I am not sure this is the expectation from the users. The image is verified by `verifier2` already. It is also verified by `verifier3`. Should it alerted if it is already verified by `verified2`?

**Review by md2119 — COMMENTED** [review]
I think some of the expectations are met, please clarify.

File: pkg/booleanpolicy/augmentedobjs/construct.go [hunk 3]
```diff
@@ -236,43 +236,22 @@ func ConstructImage(image *storage.Image) (*pathutil.AugmentedObj, error) {
...
 		for _, result := range image.GetSignatureVerificationData().GetResults() {
 			if result.GetStatus() == storage.ImageSignatureVerificationResult_VERIFIED {
-				if err := addSignatureVerificationResult(obj, i, result.GetVerifierId()); err != nil {
-					return nil, utils.Should(err)
-				}
-				i++
+				ids = append(ids, result.GetVerifierId())
 			}
 		}
 		// When the object is not created, the policy will not match, but it should match.
```
Comment (dhaus67) [inline pkg/booleanpolicy/augmentedobjs/construct.go:245]: nit: Should we keep this comment? AFAICT we anyways add an empty array here.

**Review by dhaus67 — APPROVED** [review]
Tested this also already with the integration tests I have created and the behavior is as expected.
LGTM 🚀

**Comment by md2119** [comment]
There are couple of changes required if the feature expectation is  _image must be signed by **at least one** of the selected verifiers otherwise alert_.
-  Change the english of policy criteria. Currently, the language is informing user _alert on image signed by at least one of the selected verifiers_ which means to opposite default behavior.
- Forbid `AND` operator in BE and FE to disallow _image must be signed by **all** of the selected verifiers otherwise alert_
- Given this, the query builder should be,

This translates to "alert me on images which are not verified by any of the selected verifiers."
```
func ForImageSignatureVerificationStatus() QueryBuilder {
 qbf := func(group *storage.PolicyGroup) []*query.FieldQuery {
  fq := []*query.FieldQuery{{
   Field:    search.ImageSignatureVerifiedBy.String(),
   Values:   mapValues(group, nil),
   Negate:   !group.Negate,
  }}
  if group.Negate {
    fq.Operator = query.Or    //  If the incoming is negated, it means the user is requesting to alert on images verified by at least one of the selected verifiers. With the reversal of negation, the query is (A OR B).
  } else {
    fq.Operator = query.And  // If the incoming is not negated, it means the user is requesting to alert on images not signed by any of the selected verifiers. In other words, the image must be signed by at least one of the selected verifier otherwise alert. With the reversal of negation, the query is !(A AND B) = !A OR !B (image must have at least one verifier)
  }
  return fq
 }
 return queryBuilderFunc(qbf)
}

```

**Review by md2119 — CHANGES_REQUESTED** [review]
Let's make sure we are solving the right problem. Please respond to the comments.

File: pkg/booleanpolicy/default_policies_test.go (hunk 2 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/default_policies_test.go:2225]: There will be no alert for `verified_by_2_and_3` (it is excluded from the expected matches), which I'd say is expected.

**Comment by 0x656b694d** [comment]
@md2119 , if we enabled "Not" operator for the policy, it would mean to alert users for images signed by any of the selected untrusted integrations. The current logic looks to be working as expected.

If we disallow negation on the backend, the tests for it would break. I'd like to hear a confirmation that we absolutely do not want to allow for alerting on signatures verified by untrusted integrations.

Similarly, I don't see why would we disallow operators in the API. It's up to the users to use them as they want.

**Comment by md2119** [comment]
We need to ensure the expectation remains consistent between the backend and the frontend which means removing the negation support from the backend as well.

File: pkg/booleanpolicy/augmentedobjs/construct.go (hunk 1 above)
Comment (md2119) [reply pkg/booleanpolicy/augmentedobjs/construct.go:239]: So we consider it alright to let the slice be written multiple times as it expands?

File: pkg/booleanpolicy/default_policies_test.go [hunk 4]
```diff
@@ -2230,18 +2234,16 @@ func (suite *DefaultPoliciesTestSuite) TestImageVerified() {
...
-		// several verifier IDs.
-		/*{
+		{
 			values:          []string{verifier0, verifier2},
 			negate:          false,
 			expectedMatches: allImages.Difference(set.NewFrozenStringSet("verified_by_0", "verified_by_2_and_3")),
 		},
 		{
-			values:          []string{verifier3},
+			values:          []string{verifier2, verifier3},
```
Comment (md2119) [inline pkg/booleanpolicy/default_policies_test.go:2243]: why was this necessary?

File: pkg/booleanpolicy/default_policies_test.go (hunk 2 above)
Comment (md2119) [reply pkg/booleanpolicy/default_policies_test.go:2225]: thanks for clarification

**Comment by 0x656b694d** [comment]
> We need to ensure the expectation remains consistent between the backend and the frontend which means removing the negation support from the backend as well.

I'm not sure we target our UI to provide 100% of API functionality.

File: pkg/booleanpolicy/violationmessages/printer/util.go [hunk 5]
```diff
@@ -80,7 +80,7 @@ func getSingleValueFromFieldMap(f string, fieldMap map[string][]string) (string,
 	return values[0], nil
 }
 
-func executeTemplate(tpl string, values interface{}) ([]string, error) {
+func getTemplate(tpl string) (*template.Template, error) {
```
Comment (ivan-degtiarenko) [inline pkg/booleanpolicy/violationmessages/printer/util.go:83]: Why can't we use `executeTemplate` instead of `executeTemplateT`? 

From the performance perspective amount of calls to `template.New("").Parse(tpl)` is the same considering the fact that we use lazy cache here

File: pkg/booleanpolicy/validate.go [hunk 6]
```diff
@@ -58,6 +58,15 @@ func Validate(p *storage.Policy, options ...ValidateOption) error {
 	for _, section := range p.GetPolicySections() {
 		errorList.AddError(validatePolicySection(section, configuration, p.GetEventSource()))
 	}
+
+	for _, ps := range p.PolicySections {
+		for _, pg := range ps.PolicyGroups {
+			if pg.FieldName == fieldnames.ImageSignatureVerifiedBy && pg.BooleanOperator == storage.BooleanOperator_AND {
```
Comment (ivan-degtiarenko) [inline pkg/booleanpolicy/validate.go:64]: I think that section needs an explaining comment since usually we forbid OR and AND operations. We don't want anyone mistakenly to follow our lead

File: pkg/booleanpolicy/violationmessages/printer/container.go [hunk 7]
```diff
@@ -169,10 +169,33 @@ func imageSignatureVerifiedPrinter(fieldMap map[string][]string) ([]string, erro
...
+	tmpl, err := getTemplate(imageSignatureVerifiedTemplate)
+	if err != nil {
+		return nil, err
 	}
-	return executeTemplate(imageSignatureVerifiedTemplate, r)
+
+	var result []string
+	if ids, ok := fieldMap[augmentedobjs.ImageSignatureVerifiedCustomTag]; ok {
+		for _, id := range ids {
+			if id != "" && id != "<empty>" {
```
Comment (ivan-degtiarenko) [inline pkg/booleanpolicy/violationmessages/printer/container.go:181]: Can `<empty>` be returned as one of ids here?

File: pkg/booleanpolicy/augmentedobjs/construct.go (hunk 1 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/augmentedobjs/construct.go:239]: Why would that be a problem? Multiple times = a few, most probably once.

File: pkg/booleanpolicy/violationmessages/printer/container.go (hunk 7 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/violationmessages/printer/container.go:181]: Yes, this is why this check.

File: pkg/booleanpolicy/violationmessages/printer/util.go (hunk 5 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/violationmessages/printer/util.go:83]: Mini-optimization: `executeTemplate` makes a number of extra steps like searching for the template and returning a single value array. All this looks unnecessary to do in a loop.

Comment (ivan-degtiarenko) [reply pkg/booleanpolicy/violationmessages/printer/util.go:83]: That's true. However, if we keep using `executeTemplate` it will not only keep this `util` file more simple and thus more convenient for future use, but also simplify `imageSignatureVerifierPrinter`.

From what I saw while looking at CPU and memory dumps some time ago, we really don't spend much CPU or memory to generate violations. Thus difference made by this optimization will probably not affect us in any way. 

So my suggestion is to use `executeTemplate`

Comment (0x656b694d) [reply pkg/booleanpolicy/violationmessages/printer/util.go]: Ok!

File: pkg/booleanpolicy/validate.go [hunk 8]
```diff
@@ -58,6 +58,17 @@ func Validate(p *storage.Policy, options ...ValidateOption) error {
 	for _, section := range p.GetPolicySections() {
 		errorList.AddError(validatePolicySection(section, configuration, p.GetEventSource()))
 	}
+
+	// Special case for ImageSignatureVerifiedBy policy for which we don't allow
```
Comment (md2119) [inline pkg/booleanpolicy/validate.go:62]: Add forbid operator clause to field metadata registration.  That way the need for explicit validation is removed.

**Review by md2119 — CHANGES_REQUESTED** [review]

File: pkg/booleanpolicy/validate.go (hunk 8 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/validate.go:62]: Forbid operator in metadata would forbid both AND and OR.

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM modulo Mandar's comment in `validate.go`

File: pkg/booleanpolicy/validate.go (hunk 8 above)
Comment (0x656b694d) [reply pkg/booleanpolicy/validate.go]: Got it. Please check the new commit.

**Review by md2119 — COMMENTED** [review]
Lovely. Thanks for working through the changes.

One last stop, @dhaus67  I don't see the policy criteria name issue in the UI being addressed. It is pretty much the opposite of the behavior.


@dhaus67

**Comment by dhaus67** [comment]
@md2119 yeah, we have noted that down, either me or Ivan will post a PR latest tomorrow with the changes, I'd vote to do it in a separate one if that's ok for you. It's definitely something we want to do before the cut off next week.

**Comment by md2119** [comment]
> @md2119 yeah, we have noted that down, either me or Ivan will post a PR latest tomorrow with the changes, I'd vote to do it in a separate one if that's ok for you. It's definitely something we want to do before the cut off next week.

Alright, sounds good. Just making sure it does not go unattended.

File: pkg/booleanpolicy/default_policies_test.go (hunk 4 above)
Comment (md2119) [reply pkg/booleanpolicy/default_policies_test.go:2237]: I recommend adding tc for image with mutiple verifiers with some failed verifications. For example:
```
		imageWithSignatureVerificationResults(<...>, []*storage.ImageSignatureVerificationResult{{
			VerifierId: verifier2,
			Status:     storage.ImageSignatureVerificationResult_FAILED_VERIFICATION,
		}, {
			VerifierId: verifier3,
			Status:     storage.ImageSignatureVerificationResult_VERIFIED,
		}}),
```

for policy criteria,

```
1. verifier2
2. verifier2+verifier3
```

**Review by md2119 — APPROVED** [review]

**Comment by 0x656b694d** [comment]
> I recommend adding tc for image with mutiple verifiers with some failed verifications. For example:

Such cases are already there.

File: qa-tests-backend/src/test/groovy/ImageSignatureVerificationTest.groovy [hunk 9]
```diff
@@ -146,16 +144,14 @@ w9e2Azq1OYIh/pbeBMHARDrBaqqmuMR9+BfAaPAYdkNTU6f58M2zBbuL0A==
         assert unverifiableSignatureIntegrationID
         CREATED_SIGNATURE_INTEGRATIONS.put(UNVERIFIABLE, unverifiableSignatureIntegrationID)
 
-        // TODO(ROX-9996): Uncomment after we can handle multiple verification results.
-        /* Signature integration "Distroless+Tekton" which holds both distroless and tekton cosign public keys.
+        Signature integration "Distroless+Tekton" which holds both distroless and tekton cosign public keys.
```
Comment (0x656b694d) [inline qa-tests-backend/src/test/groovy/ImageSignatureVerificationTest.groovy:147]:
```suggestion
        // Signature integration "Distroless+Tekton" which holds both distroless and tekton cosign public keys.
```

File: qa-tests-backend/src/test/groovy/ImageSignatureVerificationTest.groovy [hunk 10]
```diff
@@ -26,17 +26,15 @@ class ImageSignatureVerificationTest extends BaseSpecification {
...
 
     // List of integration names used within tests.
     // NOTE: If you add a new name, make sure to add it here.
     static final private List<String> INTEGRATION_NAMES = [
             DISTROLESS,
             TEKTON,
             UNVERIFIABLE,
-            // TODO(ROX-9996): Uncomment after we can handle multiple verification results.
-            //DISTROLESS_AND_TEKTON
+            DISTROLESS_AND_TEKTON
```
Comment (0x656b694d) [inline qa-tests-backend/src/test/groovy/ImageSignatureVerificationTest.groovy:37]:
```suggestion
            DISTROLESS_AND_TEKTON,
```
